.github/workflows/wheels.yml

```diff
@@ -171,9 +171,9 @@ jobs:
 
       - name: Upload fribidi.dll
         if: "matrix.arch != 'ARM64'"
-        uses: actions/upload-artifact@v3
+        uses: actions/upload-artifact@v4
         with:
-          name: fribidi
+          name: fribidi_${{ matrix.arch }}
           path: fribidi\*
 
   sdist:
```

Comment:
May I suggest building wheels from sdist? This is what pip does — it doesn't build wheels straight from Git source, but from the tarballs. And this is what pypa/build does if you don't pass any args. Doing so will ensure that sdists are actually usable, as a smoke test.

Here's an example: https://github.com/aio-libs/yarl/actions/runs/7258064074/workflow#L124.

P.S. `cibuildwheel` accepts a path to the sdist tarball natively (in place of a directory path) so integrating this into the wheel build jobs is as easy as downloading the artifact and feeding the sdist from it into the `cibuildwheel` action.

Reply:
Oh, and that repo also features the wheel build jobs moved into a reusable workflow and called with different params.

Reply:
> May I suggest building wheels from sdist?

We're currently using the `.ci` directory in the build process, which isn't included in the sdist
https://github.com/python-pillow/Pillow/blob/41e45b5f95c5184b8b4c4712e5614bf468331e05/.github/workflows/wheels.yml#L65-L67

Sure, we could checkout the repository AND use sdist - but that doesn't sound very neat to me.

I know you said that `cibuildwheel` can simply accept an sdist path, but we're also currently adding in extra test images, which wouldn't be quite so simple when using a compressed sdist.
https://github.com/python-pillow/Pillow/blob/41e45b5f95c5184b8b4c4712e5614bf468331e05/.github/workflows/wheels.yml#L99-L103

> Oh, and that repo also features the wheel build jobs moved into a reusable workflow

If wheels.yml gets much longer, I could see that being nice, but at the moment, I don't think it's problematic.

Also, neither of these ideas depend on this PR. If you would like to see them become part of Pillow, you could create a separate PR.

Reply:
> We're currently using the `.ci` directory in the build process, which isn't included in the sdist

And this is a reason to include it. Downstreams might want to run tests using the sdist as the source.

> If wheels.yml gets much longer, I could see that being nice, but at the moment, I don't think it's problematic.

For me, it's a structural thing that helps layer components. Not necessarily related to how big the YAML files are (although, that's also important).

> Also, neither of these ideas depend on this PR. If you would like to see them become part of Pillow, you could create a separate PR.

Sure. These are mostly drive-by comments I made having seen some familiar faces and tech involved.